src/core/or/connection_edge.c

```diff
@@ -1764,6 +1771,46 @@ parse_extended_hostname(char *address, hostname_type_t *type_out)
   return false;
 }
 
+bool lookup_anon_dns_mapping(const char *anon_address, char *onion_address_out) {
+  char *file_content = NULL;
+
+  // Check if the file exists using `file_status`
+  char *dns_fname = get_datadir_fname("anons");
+  file_status_t terms_status = file_status(dns_fname);
+  if (terms_status != FN_FILE) {
+    log_notice(LD_APP,"DNS mapping file 'anons' is not found in data dir.");
+    return false;
+  }
+
+  // Read the entire file content using `read_file_to_str`
+  file_content = read_file_to_str(dns_fname, 0, NULL);
+  if (!file_content) {
+    log_notice(LD_APP,"No mapping found for %s.",anon_address);
+    return false;
+  }
+
+  // Process the file content line by line
+  char *line = strtok(file_content, "\n");
+  while (line != NULL) {
+    char anon[256];
+    char onion[HS_SERVICE_ADDR_LEN_BASE32 + 6 + 1];
+    // Parse each line into anon and onion components
+    if (sscanf(line, "%255s %62s", anon, onion) == 2) {
```

Comment:
Should we validate the content of the anons file? At least sanity checks if operators will be able to manually edit those?

Reply:
+ to kanshi's comment here. Think best to validate content of anons file.